File: GildedRoseKata/scala/src/test/scala/com/gildedrose/SulfurasSpec.scala
package com.gildedrose

import org.scalacheck.Properties
import org.scalacheck.Prop.forAll
import Generators._

class SulfurasSpec extends Properties("Sulfuras"){

  property("Sulfuras Quality never decrese") = forAll(daysGen, sulfurasGen) {
    (days: Int, sulfuras: Item) => {
      val startingQualtity = sulfuras.quality
        (0 until days).foldLeft((true, sulfuras)){
          case ((acc: Boolean, item: Item), _: Int) => {
            val result = GildedRose.updateQuality(Array(item))
            val resultCheck = (acc && result.length == 1 && result(0).quality == startingQualtity)
            (resultCheck, result(0))
          }}._1
        }
    }
  }

File: GildedRoseKata/haskell/test/SulfurasSpec.hs
module SulfurasSpec (spec) where

import GildedRose
import Item
import Generators
import Test.Hspec
import Test.QuickCheck

spec :: Spec
spec = describe "SulfurasSpec" $
  it "should never change in quality" $ property $
    forAll sulfurasGen
      (\(i, d)->
          let days = getPositive d
              expectedQuality = (valueQ . getQuality) i
              successCondition r = length r == 1 && expectedQuality == (valueQ. getQuality . head) r
              result = foldl (\acc iteration -> acc && successCondition iteration) True $ take days (iterate updateQuality [i])
          in result)
